Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,270 @@
+  · simp only [if_pos, MvPolynomial.aeval_X]
+  · simp only [Fin.succ_ne_zero, ite_false, map_add, map_smul, map_pow,
+      MvPolynomial.aeval_X, if_pos, add_assoc, ← add_smul c, add_neg_cancel, zero_smul, add_zero]
+
+/-- `T1` leads to an algebra equiv `T`.-/
+private noncomputable abbrev T := AlgEquiv.ofAlgHom (T1 f 1) (T1 f (-1))
+  (inv_pair f 1) (by convert (inv_pair f (-1)); exact (InvolutiveNeg.neg_neg 1).symm)
+
+variable {f v} in
+private lemma lt (vlt : ∀ i, v i < up) : ∀ l ∈ ofFn v, l < up := by
+  intro l h
+  rw [mem_ofFn, Set.mem_range] at h
+  obtain ⟨y, hy⟩ := h
+  exact hy ▸ vlt y
+
+private lemma r_ne (vlt : ∀ i, v i < up) (wlt : ∀ i, w i < up) (neq : v ≠ w) :
+    ∑ x : Fin (n + 1), r x * v x ≠ ∑ x : Fin (n + 1), r x * w x := by
+  intro h
+  have : ofFn v = ofFn w := by
+    apply list_eq_of_ofDigits_eq (Nat.lt_add_right f.totalDegree one_lt_two)
+      (by simp) (lt vlt) (lt wlt)
+    simp only [ofDigits_eq_sum_mapIdx, mapIdx_eq_ofFn, get_ofFn, length_ofFn,
+      Fin.coe_cast, mul_comm]
+    simpa only [sum_ofFn] using h
+  exact neq <| Finsupp.ext <| congrFun <| ofFn_inj.mp this
+
+private lemma T_monomial {a : k} (ha : a ≠ 0) : ((T f) (monomial v a)).degreeOf 0 =
+    ∑ i : Fin (n + 1), (r i) * v i := by
+  rw [← natDegree_finSuccEquiv, monomial_eq, Finsupp.prod_pow v fun a ↦ X a]
+  simp only [Fin.prod_univ_succ, Fin.sum_univ_succ, _root_.map_mul, map_prod, map_pow,
+    AlgEquiv.ofAlgHom_apply, MvPolynomial.aeval_C, MvPolynomial.aeval_X, if_pos, Fin.succ_ne_zero,
+    ite_false, one_smul, map_add, finSuccEquiv_X_zero, finSuccEquiv_X_succ, algebraMap_eq]
+  have h1 : (finSuccEquiv k n) (C a) ≠ 0 :=
+    (map_ne_zero_iff _ (AlgEquiv.injective _)).mpr ((map_ne_zero_iff _ (C_injective _ _)).mpr ha)
+  have h2 : Polynomial.X (R := (MvPolynomial (Fin n) k))^ v 0 ≠ 0 := pow_ne_zero (v 0) X_ne_zero
+  have h3 : ∀ i : Fin n, (Polynomial.C (X (R := k) i) + Polynomial.X ^ r i.succ)
+    ^ v i.succ ≠ 0 := by
+    intro i
+    apply pow_ne_zero (v i.succ) (leadingCoeff_ne_zero.mp ?_)
+    rw [add_comm, leadingCoeff_X_pow_add_C]
+    · exact one_ne_zero
+    · simp only [add_pos_iff, ofNat_pos, true_or, pow_pos]
+  rw [natDegree_mul h1 (mul_ne_zero h2 (Finset.prod_ne_zero_iff.mpr (fun i _ ↦ h3 i))),
+    natDegree_mul h2 (Finset.prod_ne_zero_iff.mpr (fun i _ ↦ h3 i)),
+    natDegree_prod _  _ (fun i _ ↦ h3 i), natDegree_finSuccEquiv, degreeOf_C]
+  simp only [natDegree_pow, zero_add]
+  have e1 : r 0 = 1 := by simp
+  rw [natDegree_X, mul_one, Fin.val_zero, pow_zero, one_mul, add_right_inj]
+  apply Finset.sum_congr rfl
+  intro i _
+  rw [add_comm (Polynomial.C (X i)), natDegree_X_pow_add_C, mul_comm]
+
+private lemma T_degree (hv : v ∈ f.support) (hw : w ∈ f.support) (neq : v ≠ w) :
+    (T f <| monomial v <| coeff v f).degreeOf 0 ≠
+    (T f <| monomial w <| coeff w f).degreeOf 0 := by
+  rw [T_monomial _ _ <| mem_support_iff.mp hv, T_monomial _ _ <| mem_support_iff.mp hw]
+  exact r_ne f v w (up_lt f v hv) (up_lt f w hw) neq
```

Comment:
```suggestion
  exact lt_of_le_of_lt ((monomial_le_degreeOf i ‹_›).trans (degreeOf_le_totalDegree f i))
    (by omega)
```
Doing this you can remove the lemma `up_lt` (I am trying to reduce the number of private lemmas).

Reply:
Sorry, the right line is
```lean
  refine r_ne f v w (fun i ↦ ?_) (fun i ↦ ?_) neq <;>
  exact lt_of_le_of_lt ((monomial_le_degreeOf i ‹_›).trans (degreeOf_le_totalDegree f i))
    (by omega)
```